scripts/ado-terraform-nagger.py

```diff
@@ -0,0 +1,399 @@
+        if message_type == "warning":
+            # Attempt to send slack message
+            log_message_slack(slack_recipient, slack_webhook_url, message)
+        if message_type == "warning":
+            logger.warning(f"##vso[task.logissue type=warning;]{message}")
+        if message_type == "error":
+            logger.error(f"##vso[task.logissue type=error;]{message}")
+            errors_detected = True
+
+
+def extract_version(text, regex):
+    """
+    This function searches for a version number in the input `text`
+    using the specified regular expression `regex`.
+    If a match is found, the version number is returned as a string.
+    If no match is found, the function returns None.
+
+    Parameters:
+        text (str): The input text to be searched.
+        regex (str): The regular expression pattern used to search for the version.
+                     (regex must include "semver" named match group)
+
+    Returns:
+        str: The version number as a string if found, otherwise None.
+    """
+    matches = re.finditer(regex, text, re.MULTILINE | re.IGNORECASE | re.VERBOSE)
+    for match in matches:
+        return match.group("semver").strip()
+    else:
+        return None
+
+
+def terraform_version_checker(terraform_version):
+    # Load config file with pre-defined versions
+    try:
+        filename = "ado-terraform-nagger-versions.json"
+        global config
+        config = json.loads(load_file(filename))
+    except json.JSONDecodeError:
+        logger.error(f"{filename} contains invalid JSON")
+    except Exception as e:
+        logger.error(f"Error loading {filename}: {e}")
+
+    # Error if terraform version is lower than specified within config.
+    if version.parse(terraform_version) < version.parse(
+        config["terraform"]["error_below"]
+    ):
+        log_message(
+            slack_user_id,
+            slack_webhook_url,
+            "error",
+            f"Detected terraform version {terraform_version} "
+            f'is lower than {config["terraform"]["error_below"]}. '
+            "This is no longer allowed, please upgrade...",
+        )
+
+    # Warn if terraform version is lower than specified within config.
+    if version.parse(terraform_version) < version.parse(
+        config["terraform"]["warn_below"]
+    ):
+        log_message(
+            slack_user_id,
+            slack_webhook_url,
+            "warning",
+            f"Detected terraform version {terraform_version} "
+            f'is lower than {config["terraform"]["warn_below"]}. '
+            "Please upgrade...",
+        )
+
+
+def main():
+
+    # Retrieve HMCTS github to slack user mappings
+    hmcts_github_slack_user_mappings = get_hmcts_github_slack_user_mappings()
+    # Attempt to retrieve github username
+    github_user = os.getenv("BUILD_SOURCEVERSIONAUTHOR")
+    # Attempt to map github user to slack username
+    global slack_user_id
+    slack_user_id = get_github_slack_user_mapping(
```

Comment:
(optional for spike, but required for feature completeness)
this should have different behaviour on master / PR. i.e. on PR it should notify the main build notices channel